raster/src/main/scala/geotrellis/raster/BufferTile.scala

```diff
@@ -0,0 +1,441 @@
+    val bufferLeft = math.min(this.bufferLeft, other.bufferLeft)
+    val bufferRight = math.min(this.bufferRight, other.bufferRight)
+    val bufferBottom = math.min(this.bufferBottom, other.bufferBottom)
+    val cols = bufferLeft + gridBounds.width + bufferRight
+    val rows = bufferTop + gridBounds.height + bufferBottom
+
+    val tile = ArrayTile.alloc(cellType.union(other.cellType), cols, rows)
+
+    // index both tiles relative to (0, 0) pixel
+    cfor(-bufferTop)(_ < gridBounds.height + bufferRight, _ + 1) { row =>
+      cfor(-bufferLeft)(_ < gridBounds.width + bufferRight, _ + 1) { col =>
+        val leftV = this.getDouble(col, row)
+        val rightV = other.getDouble(col, row)
+        tile.setDouble(col + bufferLeft, row + bufferTop, f(leftV, rightV))
+      }
+    }
+
+    if (bufferTop + bufferLeft + bufferRight + bufferBottom == 0) tile
+    else BufferTile(tile, GridBounds[Int](
+      colMin = bufferLeft,
+      rowMin = bufferTop,
+      colMax = bufferLeft + gridBounds.width - 1,
+      rowMax = bufferTop + gridBounds.height - 1)
+    )
+  }
+
+  /**
+    * Combine two tiles' cells into new cells using the given integer
+    * function. For every (x, y) cell coordinate, get each of the
+    * tiles' integer values, map them to a new value, and assign it to
+    * the output's (x, y) cell.
+    *
+    * @param   other  The other Tile
+    * @param   f      A function from (Int, Int) to Int
+    * @return         The result, an Tile
+    */
+  def combine(other: Tile)(f: (Int, Int) => Int): Tile = {
+    (this, other).assertEqualDimensions()
+
+    other match {
+      case bt: BufferTile =>
+        this.combine(bt)(f)
+      case _ =>
+        val tile = ArrayTile.alloc(cellType.union(other.cellType), cols, rows)
+        cfor(0)(_ < rows, _ + 1) { row =>
+          cfor(0)(_ < cols, _ + 1) { col =>
+            tile.set(col, row, f(get(col, row), other.get(col, row)))
+          }
+        }
+        tile
+    }
+  }
+
+  /**
+    * Combine two tiles' cells into new cells using the given double
+    * function. For every (x, y) cell coordinate, get each of the
+    * tiles' double values, map them to a new value, and assign it to
+    * the output's (x, y) cell.
+    *
+    * @param   other  The other Tile
+    * @param   f      A function from (Int, Int) to Int
+    * @return         The result, an Tile
+    */
+  def combineDouble(other: Tile)(f: (Double, Double) => Double): Tile = {
+    (this, other).assertEqualDimensions()
+
+    other match {
+      case bt: BufferTile =>
+        this.combineDouble(bt)(f)
+      case _ =>
+        val tile = ArrayTile.alloc(cellType, cols, rows)
+        cfor(0)(_ < rows, _ + 1) { row =>
+          cfor(0)(_ < cols, _ + 1) { col =>
+            tile.setDouble(col, row, f(getDouble(col, row), other.getDouble(col, row)))
+          }
+        }
+        tile
+    }
+  }
+
+  def mapTile(f: Tile => Tile): BufferTile = BufferTile(f(sourceTile), gridBounds)
```

Comment:
@echeipesh how about making all functions to work within the window, but if user wants to apply smth beayond the window they need to do it explicitly. Mb the function name could be different though.